app/src/main/java/org/mozilla/fenix/perf/StartupManager.kt

```diff
@@ -0,0 +1,280 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.perf
+
+import android.app.Activity
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.widget.LinearLayout
+import androidx.annotation.VisibleForTesting
+import androidx.navigation.NavDestination
+import androidx.navigation.fragment.NavHostFragment
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.launch
+import mozilla.components.browser.session.SessionManager
+import mozilla.components.feature.privatemode.notification.PrivateNotificationFeature
+import mozilla.components.support.ktx.android.arch.lifecycle.addObservers
+import mozilla.components.support.utils.SafeIntent
+import mozilla.components.support.utils.toSafeIntent
+import mozilla.components.support.webextensions.WebExtensionPopupFeature
+import org.mozilla.fenix.GleanMetrics.Metrics
+import org.mozilla.fenix.HomeActivity
+import org.mozilla.fenix.R
+import org.mozilla.fenix.browser.UriOpenedObserver
+import org.mozilla.fenix.browser.browsingmode.BrowsingMode
+import org.mozilla.fenix.browser.browsingmode.BrowsingModeManager
+import org.mozilla.fenix.browser.browsingmode.DefaultBrowsingModeManager
+import org.mozilla.fenix.components.metrics.BreadcrumbsRecorder
+import org.mozilla.fenix.components.metrics.Event
+import org.mozilla.fenix.customtabs.ExternalAppBrowserActivity
+import org.mozilla.fenix.ext.breadcrumb
+import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.ext.settings
+import org.mozilla.fenix.home.intent.HomeIntentProcessor
+import org.mozilla.fenix.home.intent.StartSearchIntentProcessor
+import org.mozilla.fenix.session.PrivateNotificationService
+import org.mozilla.fenix.theme.DefaultThemeManager
+import org.mozilla.fenix.theme.ThemeManager
+import java.lang.ref.WeakReference
+import kotlin.reflect.KProperty0
+
+class StartupManager {
+
+    @Suppress("TooManyFunctions")
+    companion object {
+        @OptIn(ExperimentalCoroutinesApi::class)
+        @Suppress("LongParameterList", "LongMethod")
+        fun activityOnCreate(
+            activity: HomeActivity,
+            savedInstanceState: Bundle?,
+            isVisuallyComplete: Boolean,
+            webExtensionPopupFeature: WebExtensionPopupFeature,
+            getBreadcrumbMessage: (NavDestination) -> String,
+            externalSourceIntentProcessors: KProperty0<List<HomeIntentProcessor>>,
+            action: () -> Unit
+        ): StartupData {
+            var themeManager: ThemeManager?
+            var browsingModeManager: BrowsingModeManager?
```

Comment:
nit: don't hoist variables: declare them close to where you need them. In that case, these can probably be `val` too